### PR TITLE
Mpcdi Support

### PR DESCRIPTION
**Pull request by GPayne**
This branch was used for developing support for the MPCDI format, so that SGCT can be used on the Evans & Sutherland Digistar projector system (and others that support this format).
I merged master into this branch prior to the pull request (including recent commit for OpenGL 4.6 support), built in Release and tested by running all 31 examples that are created when CMake options SGCT_EXAMPLES and SGCT_EXAMPLES_IMGUI are enabled. All examples ran without error.
I created 2 additional test cases that use mpcdi support in the 'calibrator' example. The directory has 2 batch files that run these mpcdi cases (with additional config files included there).

### CODE REVIEW

File: include/sgct/SGCTMpcdi.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
+/*
+ * SGCTMpcdi.h
+ *
+ *  Created on: Jul 3, 2017
+ *      Author: openspace
```
Comment (alexanderbock) [inline include/sgct/SGCTMpcdi.h]: You can put your own name there, too

File: include/sgct/SGCTMpcdi.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#ifndef SGCT_DONT_USE_EXTERNAL
+    #include <external/tinyxml2.h>
+#else
+    #include <tinyxml2.h>
+#endif
+
+namespace sgct_core //simple graphics cluster toolkit
+{
+
+struct mpcdiSubFiles {
```
Comment (alexanderbock) [inline include/sgct/SGCTMpcdi.h]: Capitalize the first character

File: include/sgct/SGCTMpcdi.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    ~mpcdiSubFiles() {
+        for (int i = 0; i < mpcdi_nRequiredFiles; ++i) {
+            if( buffer[i] != nullptr )
+                delete buffer[i];
+        }
+    }
+};
+
+struct mpcdiRegion {
```
Comment (alexanderbock) [inline include/sgct/SGCTMpcdi.h]: Capitalize first letter

File: include/sgct/SGCTMpcdi.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    std::string id;
+};
+
+struct mpcdiWarp {
```
Comment (alexanderbock) [inline include/sgct/SGCTMpcdi.h]: Capitalize first letter

File: include/sgct/SGCTMpcdi.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    std::string id;
+    std::string pathWarpFile;
+    bool haveFoundPath = false;
+    bool haveFoundInterpolation = false;
+};
+
+struct mpcdiFoundItems {
```
Comment (alexanderbock) [inline include/sgct/SGCTMpcdi.h]: Capitalize first letter

File: include/sgct/Viewport.h [hunk 6]
```diff
@@ -16,9 +16,38 @@ For conditions of distribution and use, see copyright notice in sgct.h
...
+    #include <external/tinyxml2.h>
+#else
+    #include <tinyxml2.h>
+#endif
+#define MAX_XML_DEPTH 16
+
+
 namespace sgct_core
 {
+struct frustumData {
```
Comment (alexanderbock) [inline include/sgct/Viewport.h]: Capitalize first letter

File: lib/mingw_x64/merge.bat [hunk 7]
```diff
@@ -10,4 +11,4 @@ ar x deps/libvrpn.a
 ar x libsgct_light.a
 ar r libsgct.a *.obj
 DEL *.o *.obj
-pause
\ No newline at end of file
+pause
```
Comment (alexanderbock) [inline lib/mingw_x64/merge.bat]: Why the `pause`?

File: lib/mingw_x64/merge_debug.bat [hunk 8]
```diff
@@ -10,4 +11,4 @@ ar x deps/libvrpnd.a
 ar x libsgct_lightd.a
 ar r libsgctd.a *.obj
 DEL *.o *.obj
-pause
\ No newline at end of file
+pause
```
Comment (alexanderbock) [inline lib/mingw_x64/merge_debug.bat]: Why the `pause`?

File: src/apps/calibrator/run_mpcdi_example1.bat [hunk 9]
```diff
@@ -0,0 +1 @@
+calibrator.exe --Ignore-Sync -tilt 0.0 -radius 5.5 -config mpcdi_example1_distorted.xml
```
Comment (alexanderbock) [inline src/apps/calibrator/run_mpcdi_example1.bat]: Missing new line at the end of the file

File: src/apps/calibrator/run_mpcdi_example2.bat [hunk 10]
```diff
@@ -0,0 +1 @@
+calibrator.exe --Ignore-Sync -tilt 0.0 -radius 5.5 -config mpcdi_example2_undistorted.xml
```
Comment (alexanderbock) [inline src/apps/calibrator/run_mpcdi_example2.bat]: Missing new line at the end of the file

File: src/sgct/SGCTMpcdi.cpp [hunk 11]
```diff
@@ -0,0 +1,517 @@
+/*
+ * SGCTMpcdi.cpp
+ *
+ *  Created on: Jul 3, 2017
+ *      Author: openspace
```
Comment (alexanderbock) [inline src/sgct/SGCTMpcdi.cpp]: You can put your name in here

File: src/sgct/SGCTMpcdi.cpp [hunk 12]
```diff
@@ -0,0 +1,517 @@
...
+#include <sgct/ClusterManager.h>
+
+#include <sgct/SGCTSettings.h>
+#include <sgct/SGCTMpcdi.h>
+#include <algorithm>
+#include <sstream>
+#include "unzip.h"
+#include <zip.h>
+
+sgct_core::SGCTMpcdi::SGCTMpcdi(std::string& parentErrorMessage) : mErrorMsg(parentErrorMessage)
```
Comment (alexanderbock) [inline src/sgct/SGCTMpcdi.cpp]: This string can be non-ref and then use `std::move(parentErrorMessage)`

File: src/sgct/SGCTMpcdi.cpp [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+{
+}
+
+sgct_core::SGCTMpcdi::~SGCTMpcdi()
+{
+    for (mpcdiWarp* ptr : mWarp)
+    {
+        if (ptr)
```
Comment (alexanderbock) [inline src/sgct/SGCTMpcdi.cpp]: The `if` check is superflous.  `T* t = nullptr; delete t;` is fine

File: src/sgct/SGCTMpcdi.cpp [hunk 14]
```diff
@@ -0,0 +1,517 @@
...
+
+bool sgct_core::SGCTMpcdi::parseConfiguration(const std::string filenameMpcdi,
+	SGCTNode& tmpNode,
+	sgct::SGCTWindow& tmpWin)
+{
+	FILE * cfgFile = nullptr;
+	unzFile zipfile;
+	const int MaxFilenameSize_bytes = 500;
+
+    if (!openZipFile(cfgFile, filenameMpcdi, &zipfile))
```
Comment (alexanderbock) [inline src/sgct/SGCTMpcdi.cpp]: Cleaner to move all of these outside of the `if` statement and then check the `bool` explicitly

Comment (alexanderbock) [reply src/sgct/SGCTMpcdi.cpp]: Same for the ones below

File: src/sgct/Viewport.cpp [hunk 15]
```diff
@@ -50,6 +51,9 @@ sgct_core::Viewport::~Viewport()
 
     if (mBlackLevelMaskTextureIndex)
         glDeleteTextures(1, &mBlackLevelMaskTextureIndex);
+
+    if (mMpcdiWarpMeshData)
```
Comment (alexanderbock) [inline src/sgct/Viewport.cpp]: No need for this `if` check

File: include/sgct/SGCTMpcdi.h (hunk 2 above)
Comment (GPayne) [reply include/sgct/SGCTMpcdi.h]: Done

File: include/sgct/SGCTMpcdi.h (hunk 1 above)
Comment (GPayne) [reply include/sgct/SGCTMpcdi.h]: Done

File: include/sgct/SGCTMpcdi.h (hunk 3 above)
Comment (GPayne) [reply include/sgct/SGCTMpcdi.h]: Done

File: include/sgct/SGCTMpcdi.h (hunk 4 above)
Comment (GPayne) [reply include/sgct/SGCTMpcdi.h]: Done

File: include/sgct/SGCTMpcdi.h (hunk 5 above)
Comment (GPayne) [reply include/sgct/SGCTMpcdi.h]: Done

File: include/sgct/Viewport.h (hunk 6 above)
Comment (GPayne) [reply include/sgct/Viewport.h]: Done

File: lib/mingw_x64/merge.bat (hunk 7 above)
Comment (GPayne) [reply lib/mingw_x64/merge.bat]: I don't know on this one. I removed it.

File: lib/mingw_x64/merge_debug.bat (hunk 8 above)
Comment (GPayne) [reply lib/mingw_x64/merge_debug.bat]: Removed this one also.

File: src/apps/calibrator/run_mpcdi_example1.bat (hunk 9 above)
Comment (GPayne) [reply src/apps/calibrator/run_mpcdi_example1.bat]: Added a new line

File: src/apps/calibrator/run_mpcdi_example2.bat (hunk 10 above)
Comment (GPayne) [reply src/apps/calibrator/run_mpcdi_example2.bat]: Added a new line

File: src/sgct/SGCTMpcdi.cpp (hunk 11 above)
Comment (GPayne) [reply src/sgct/SGCTMpcdi.cpp]: Done

File: src/sgct/SGCTMpcdi.cpp (hunk 13 above)
Comment (GPayne) [reply src/sgct/SGCTMpcdi.cpp]: So what you're saying is that either the `ptr` value in this for loop will be a valid address or `nullptr`. If that's the case then I agree and will remove the `if`.

Comment (alexanderbock) [reply src/sgct/SGCTMpcdi.cpp]: The check `if (ptr)` just checks whether `ptr` is `nullptr` or a valid pointer.  In either case, calling `delete` on it is totally fine . `T* t = nullptr; delete t;` is a null-op.

File: src/sgct/SGCTMpcdi.cpp (hunk 14 above)
Comment (GPayne) [reply src/sgct/SGCTMpcdi.cpp]: Did this on a number of the function calls

File: src/sgct/Viewport.cpp (hunk 15 above)
Comment (GPayne) [reply src/sgct/Viewport.cpp]: Done

**Pull request merged**